source/common/protobuf/utility.cc

```diff
@@ -136,18 +129,31 @@ class ApiBoostRetryException : public EnvoyException {
 // This relies on the property that any v3 configuration that is readable as
 // v2 has the same semantics in v2/v3, which holds due to the highly structured
 // vN/v(N+1) mechanical transforms.
-void tryWithApiBoosting(MessageXformFn f, Protobuf::Message& message) {
+void tryWithApiBoosting(MessageXformFn f, Protobuf::Message& message,
+                        absl::optional<MessageVersion> version) {
   const Protobuf::Descriptor* earlier_version_desc =
       Config::ApiTypeOracle::getEarlierVersionDescriptor(message.GetDescriptor()->full_name());
-  // If there is no earlier version of a message, just apply f directly.
-  if (earlier_version_desc == nullptr) {
+  // If there is no earlier version of a message or we only accept the latest version, just apply f
+  // directly.
+  if (earlier_version_desc == nullptr || version == MessageVersion::LATEST_VERSION) {
     f(message, MessageVersion::LATEST_VERSION);
     return;
   }
+
   Protobuf::DynamicMessageFactory dmf;
-  auto earlier_message = ProtobufTypes::MessagePtr(dmf.GetPrototype(earlier_version_desc)->New());
-  ASSERT(earlier_message != nullptr);
+  // Only try the earlier version if version is specified as such.
+  if (version == MessageVersion::EARLIER_VERSION) {
+    auto earlier_message = ProtobufTypes::MessagePtr(dmf.GetPrototype(earlier_version_desc)->New());
+    ASSERT(earlier_message != nullptr);
+    f(*earlier_message, MessageVersion::EARLIER_VERSION);
+    Config::VersionConverter::upgrade(*earlier_message, message);
```

Comment:
The repetition with the code below might support refactoring if we go down this path.